dts/bindings/gpio/gpio-leds.yaml

```diff
@@ -15,3 +22,7 @@ child-binding:
           required: true
           type: string
           description: Human readable string describing the device (used as device_get_binding() argument)
+       id:
```

Comment:
This is an API change, incompatible with Linux where `gpio-leds` is defined.

Reply:
Also the example devicetree modifications use identifiers that start with 1, which is unusual.  Is `id` intended to be an opaque label, or an ordinal identifying the LED based on position in some ordering?

Finally the Zephyr `led` API requires this to be interpreted as an unsigned value, which should be noted (if this property is acceptable in this compatible).

Reply:
Why we need to be compatible with the Linux ? I do no have to start from 1, I did this by chance. ID is used to identify the LED by API, because the GPIO port can be different for every LEDs in the `leds` node so the LEDs can have the same GPIO Pin so we cannot drive LEDs using it. That is why I provide a ID property which only should be unique for each LED. ID is used in the LED API to control the leds for_example `led_on(led, id)`. Do you have any idea how to get ride with this ID property without changing whole led API and maintaining compatibility with Linux ?

Reply:
I don't believe there's any guarantee that the order of children is preserved across builds, but that would probably be least intrusive way to go.

Reply:
Arf. I am always forgetting this one.

I think it is possible to make this index/ID "optional" and then to keep compatibility with the existing DTS coming from Linux. If the property is available in the LED subnode, then the led-gpio driver could use it. This would allow to ensure the LED order. But if the property is not available then the index of the LED in the array could be used. [26101](https://github.com/zephyrproject-rtos/zephyr/pull/26101) introduces a led_get_info syscall. This syscall allows to retrieve information (label, index, ...) for a given LED number. If this callback is provided by the led_gpio driver then a user would still be able to discover which LED label is mapped to which LED number (if the ID to enforce it is missing).

And to finish, if an ID/index is required, then obviously I would like to see the index led-controller property from [26101](https://github.com/zephyrproject-rtos/zephyr/pull/26101) used :)

Reply:
At this moment this ID property is optional but I assumed that if you not defining it in dts, you should not use this driver. Since we do not have this `led_get_info` functionality removing id and basing on index which depends on the leds order in board dts is a little risky like @pabigot said there is not guarantee that somebody do not change the leds order in dts. This risk is probably very small.

Reply:
Well, the "id" property is required to use the led-gpio driver. Without this property, the driver won't work. So it should not be marked as optional in the binding because it is not. But if you allow the driver to work even if the "id" is not provided (by using indexes) then the "id" property becomes truly optional. And the driver becomes compatible with the existing board DTS from Linux as well.

Anyway, I think it would be nice to have this PR based on [26101](https://github.com/zephyrproject-rtos/zephyr/pull/26101). This would provide the led_get_info syscall and the index property.

Reply:
> Why we need to be compatible with the Linux ?

This is a documented design goal:

https://docs.zephyrproject.org/latest/guides/dts/design.html#source-compatibility-with-other-operating-systems

> I don't believe there's any guarantee that the order of children is preserved across builds, but that would probably be least intrusive way to go.

What if we just added a new devicetree API macro named `DT_CHILD_INDEX(node_id)` which let you get the index of `node_id` in its sibling list?

So for example, if this is the dts:

```
        leds {
                compatible = "gpio-leds";
                red_led: led_0 {
                        gpios = < &gpiob 0x16 0x1 >;
                        label = "User LD1";
                };
                green_led: led_1 {
                        gpios = < &gpioe 0x1a 0x1 >;
                        label = "User LD2";
                };
                blue_led: led_2 {
                        gpios = < &gpiob 0x15 0x1 >;
                        label = "User LD3";
                };
        };
```

then:

- `DT_CHILD_INDEX(DT_NODELABEL(red_led))` is 0
- `DT_CHILD_INDEX(DT_NODELABEL(green_led))` is 1
- `DT_CHILD_INDEX(DT_NODELABEL(blue_led))` is 2

If I understand correctly, that at least is the necessary mechanism.

Or do people just want to wait for and rebase on top of #26101?

Reply:
> . The index in the `leds[]` array provides the id that identifies that LED.

I see how that takes you from a parent node an array initializer.

What I don't follow is how to start with a devicetree node referring to a child (i.e. referring to an LED), and get an index.

Reply:
I guess it depends on how this is supposed to be used.  But you're right, if you want to be able to identify a specific node by its devicetree name, then `DT_CHILD_INDEX()` is the way to do it.